build.gradle.kts

```diff
@@ -0,0 +1,105 @@
+import org.gradle.api.tasks.Delete
+import org.gradle.api.tasks.JavaExec
+import org.gradle.kotlin.dsl.getValue
+import org.gradle.kotlin.dsl.kotlin
+import org.gradle.kotlin.dsl.maven
+import org.gradle.kotlin.dsl.repositories
+import org.gradle.api.plugins.quality.CheckstyleExtension
+
+import com.android.build.gradle.BaseExtension
+import org.jetbrains.kotlin.gradle.dsl.Coroutines
+import org.jetbrains.kotlin.gradle.dsl.KotlinProjectExtension
+
+buildscript {
+    repositories {
+        google()
+        jcenter()
+        maven {
+            url = uri("https://maven.fabric.io/public")
+        }
+    }
+    dependencies {
+        classpath("com.android.tools.build:gradle:3.0.1")
+        classpath("io.fabric.tools:gradle:1.25.1")
+        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${deps.versions.kotlin}")
+    }
+}
+
+plugins {
+    id("com.github.ben-manes.versions") version "0.17.0"
+}
+
+allprojects {
+    repositories {
+        google()
+        jcenter()
+        mavenLocal()
+        mavenCentral()
+    }
+}
+
+subprojects {
+    apply {
+        plugin("checkstyle")
+    }
+
+    extensions.configure(CheckstyleExtension::class.java) {
+        configFile = file("config/checkstyle/checkstyle.xml")
+        isIgnoreFailures = false
+        toolVersion = "6.17"
+    }
+
+    afterEvaluate {
+        // BaseExtension is common parent for application, library and test modules
+        extensions.configure(BaseExtension::class.java) {
+            compileSdkVersion(deps.android.compileSdkVersion)
+            buildToolsVersion(deps.android.buildToolsVersion)
+            defaultConfig {
+                minSdkVersion(deps.android.minSdkVersion)
+                targetSdkVersion(deps.android.targetSdkVersion)
+                multiDexEnabled = true
+            }
+            lintOptions {
+                isAbortOnError = true
+                disable("UnusedResources") // https://issuetracker.google.com/issues/63150366
+                disable("InvalidPackage")
+            }
+            dexOptions {
+                dexInProcess = true
+            }
+            compileOptions {
+                setSourceCompatibility(JavaVersion.VERSION_1_8)
+                setTargetCompatibility(JavaVersion.VERSION_1_8)
+            }
+        }
+
+        extensions.configure(KotlinProjectExtension::class.java) {
+            experimental.coroutines = Coroutines.ENABLE
+        }
+    }
+
+    configurations {
+        all {
+            exclude(group = "com.google.code.findbugs", module = "jsr305")
+        }
+    }
+}
+
+configurations {
+    maybeCreate("ktlint")
+}
+
+dependencies {
+    "ktlint"("com.github.shyiko:ktlint:0.14.0")
+}
+
+tasks {
+    "clean"(Delete::class) {
```

Comment:
dafuq kind of syntax is this. I hope this extension method is just inside the tasks scope...

Reply:
Looks weird, borrowed it from few examples in gradle repo. As soon as there is no well-established gradle-kotlin-script guidelines, I'd stick with this approach. Looks somewhat groovyish.

Yes, it's inside tasks scope — https://github.com/gradle/kotlin-dsl/blob/master/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt

Reply:
Is this really the recommended syntax?

Reply:
There is no well-established syntax yet, but latest samples are intensively using this syntax: https://github.com/gradle/kotlin-dsl/blob/a5de7d1b485ae30755d4ac09e54063cd9e9678e3/samples/task-dependencies/build.gradle.kts